### ISSUE TITLE
Timing of effects

### ISSUE THREAD

**Issue by thysultan** [reactjs/reactjs.org#1598]
>Although `useEffect` is deferred until after the browser has painted, it's guaranteed to fire before any new renders. React will always flush a previous render's effects before starting a new update.

Is this true in both  concurrent and non-concurrent mode. It would appear intuitively that these two call-sites in non-concurrent mode would be deterministic:

```js
ReactDOM.render(<Foo>, container)
ReactDOM.render(<Foo>, container)
```

That is it would invoke a new render before the effect in the first render has been invoked. 

If so would it make more sense to specify it is only in relation to effects i.e previous effects are guaranteed to run before future effects.

As it stands, it looks appears as though the "render" call-site is non-deterministic in non-concurrent.
